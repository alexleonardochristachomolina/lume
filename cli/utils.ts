import lume from "../mod.ts";
import { toFileUrl } from "../deps/path.ts";
import { isUrl } from "../core/utils/path.ts";
import { getConfigFile } from "../core/utils/lume_config.ts";
import { log } from "../core/utils/log.ts";
import Site from "../core/site.ts";

/** Create a site instance */
export async function createSite(config?: string): Promise<Site> {
  let url: string | undefined;
  const watchInclude = [];

  if (config && isUrl(config)) {
    url = config;
  } else {
    const path = await getConfigFile(config);

    if (path) {
      watchInclude.push(path);
      url = toFileUrl(path).href;
    }
  }

  if (url) {
    log.info(`Loading config file <gray>${url}</gray>`);
    const mod = await import(url);
    const site = mod.default as Site | undefined;
    if (!(site instanceof Site)) {
      log.critical(
        `[Lume] Missing Site instance! Ensure your config file does export the Site instance as default.`,
      );
      throw new Error("Site instance is not found");
    }

    site._data.configFile = url;
    site.options.watcher.include.push(...watchInclude);
    return site;
  }

  return lume();
}

/** Create a site intance and build it */
export async function buildSite(config?: string): Promise<Site> {
  const site = await createSite(config);

  performance.mark("start");
  await site.build();
  performance.mark("end");

  log.info(`üçæ Site built into <gray>${site.options.dest}</gray>`);
  const duration = performance.measure("duration", "start", "end").duration /
    1000;
  const total = site.pages.length + site.files.length;
  log.info(
    `  <gray>${total} files generated in ${duration.toFixed(2)} seconds</gray>`,
  );

  return site;
}
