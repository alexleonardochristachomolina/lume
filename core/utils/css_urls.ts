type UrlWalker = (url: string, type: string) => string | Promise<string>;
type Position = [string, string, number, number];
type Index = [string, number];

export async function walkUrls(
  code: string,
  walker: UrlWalker,
): Promise<string> {
  let result = "";
  let index = 0;
  let position = next(code, index);

  while (position) {
    const [type, url, start, end] = position;
    const newUrl = await walker(url, type);
    result += `${code.slice(index, start)}"${newUrl}"`;
    index = end;
    position = next(code, index);
  }

  return result + code.slice(index);
}

function next(code: string, index = 0): Position | undefined {
  const nextUrl = code.indexOf("url(", index);
  const nextComment = code.indexOf("/*", index);
  const nextImport = code.indexOf("@import ", index);

  let found: Index | undefined;

  if (nextUrl !== -1) {
    found = ["url", nextUrl];
  }
  if (nextComment !== -1 && (!found || nextComment < found[1])) {
    found = ["comment", nextComment];
  }
  if (nextImport !== -1 && (!found || nextImport < found[1])) {
    found = ["import", nextImport];
  }

  if (!found) {
    return;
  }

  // If it's a comment, find the end of it
  if (found[0] === "comment") {
    const end = code.indexOf("*/", found[1]);
    if (end === -1) {
      return;
    }
    return next(code, end + 2);
  }

  if (found[0] === "import") {
    const double = code.indexOf('"', found[1]);
    const single = code.indexOf("'", found[1]);
    const urlFn = code.indexOf("url(", found[1]);

    let foundImport: Index | undefined;

    if (double !== -1) {
      foundImport = ["double", double];
    }
    if (single !== -1 && (!foundImport || single < foundImport[1])) {
      foundImport = ["single", single];
    }

    // It's has a url(), so delegate to the url() logic
    if (urlFn !== -1 && (!foundImport || urlFn < foundImport[1])) {
      return next(code, urlFn);
    }

    if (!foundImport) {
      return;
    }

    const start = foundImport[1];
    const end =
      code.indexOf(foundImport[0] === "double" ? '"' : "'", start + 1) + 1;
    return ["url", code.slice(start + 1, end - 1), start, end];
  }

  if (found[0] === "url") {
    const start = code.indexOf("(", found[1]) + 1;
    const end = code.indexOf(")", start);
    let url = code.slice(start, end).trim();
    // Remove quotes
    if (url.startsWith('"') || url.startsWith("'")) {
      url = url.slice(1, -1);
    }
    return ["url", url, start, end];
  }
}
