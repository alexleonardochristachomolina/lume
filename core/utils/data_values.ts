import type { Data } from "../file.ts";

/**
 * Get the value of a page data
 * For example, if the value is "=title", it returns the value of the page data "title"
 * If the value is "$.title", it will return the value of the element with the selector ".title"
 */
export function getDataValue(data: Partial<Data>, value?: unknown) {
  // Get the value from the page data
  if (typeof value === "string") {
    return searchValue(data, value);
  }

  if (typeof value === "function") {
    return value(data);
  }

  return value;
}

function searchValue(data: Partial<Data>, value: string): unknown {
  if (!value) {
    return;
  }

  if (value.startsWith("=")) {
    let key = value.slice(1);
    [key, value] = parseFallback(key);

    if (!key.includes(".")) {
      return data[key] ?? searchValue(data, value);
    }

    const keys = key.split(".");
    let val = data;
    for (const key of keys) {
      val = val?.[key];
    }
    return val ?? searchValue(data, value);
  }

  if (value.startsWith("$")) {
    let selector = value.slice(1);
    [selector, value] = parseFallback(selector);

    return queryCss(selector, data.page?.document) ?? searchValue(data, value);
  }

  return value;
}

function parseFallback(key: string): [string, string] {
  const fallback = key.indexOf("||");

  if (fallback !== -1) {
    return [
      key.slice(0, fallback).trim(),
      key.slice(fallback + 2).trim(),
    ];
  }

  return [
    key,
    "",
  ];
}

// https://regexr.com/7qnot
const checkForAttrPattern = /^(.+)\s+(?:attr\(([\w\-]+)\))$/;

function queryCss(query: string, document?: Document) {
  const checkResult = query.match(checkForAttrPattern);

  const hasAttr = checkResult?.[0];
  if (hasAttr) {
    const [_, query, name] = checkResult;
    return document?.querySelector(query)?.getAttribute(name);
  }

  return document?.querySelector(query)?.innerHTML;
}
