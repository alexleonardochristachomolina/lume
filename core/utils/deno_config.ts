import { parse } from "../../deps/jsonc.ts";
import { isUrl } from "../utils/path.ts";

/** Import map file */
export interface ImportMap {
  imports: Record<string, string>;
  scopes?: Record<string, Record<string, string>>;
}

/** Basic options for deno.json file */
export interface DenoConfig extends Partial<ImportMap> {
  importMap?: string;
  tasks?: Record<string, string>;
  compilerOptions?: {
    jsx?: "jsx" | "react-jsx" | "precompile";
    jsxImportSource?: string;
    types?: string[];
  };
  [key: string]: unknown;
}

export interface DenoConfigResult {
  file: string;
  config: DenoConfig;
  importMap?: ImportMap;
}

/** Detect and returns the Deno configuration */
export async function readDenoConfig(): Promise<DenoConfigResult | undefined> {
  for (const file of ["deno.json", "deno.jsonc"]) {
    try {
      const content = await Deno.readTextFile(file);
      const config = parse(content) as DenoConfig;
      let importMap: ImportMap | undefined;

      if (config.importMap) {
        importMap = isUrl(config.importMap)
          ? await (await fetch(config.importMap)).json()
          : await JSON.parse(await Deno.readTextFile(config.importMap));
      } else if (config.imports) {
        importMap = {
          imports: config.imports,
          scopes: config.scopes,
        };
      }
      return { file, config, importMap };
    } catch (err) {
      if (err instanceof Deno.errors.NotFound) {
        continue;
      }

      throw err;
    }
  }
}
