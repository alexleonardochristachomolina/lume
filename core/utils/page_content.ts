import type { Page } from "../file.ts";

export function insertContent(
  page: Page,
  content: string,
  placeholder?: string,
) {
  const pageContent = page.content;

  if (pageContent && typeof pageContent === "string") {
    if (placeholder && pageContent.includes(placeholder)) {
      page.content = pageContent.replace(
        placeholder,
        content,
      );
    } else {
      page.content += `\n${content}`;
    }
  } else {
    page.content = content;
  }
}
