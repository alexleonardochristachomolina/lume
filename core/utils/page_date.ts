import { getGitDate, parseDate } from "./date.ts";

import type { Page, RawData } from "../file.ts";

/** Returns the Date instance of a file */
export function getPageDate(page: Page): Date {
  const data = page.data as RawData;
  const { date } = data;

  if (date instanceof Date) {
    return date;
  }

  if (typeof date === "number") {
    return new Date(date);
  }

  const { entry } = page.src;
  const info = entry?.getInfo();

  if (typeof date === "string") {
    if (entry && info) {
      switch (date.toLowerCase()) {
        case "git last modified":
          return getGitDate("modified", entry.src) || info.mtime || new Date();
        case "git created":
          // getGitDate("created", ...) uses `git log --diff-filter=A`
          // which returns nothing for files where the latest commit had a merge conflict.
          // Therefore we fallback to last modified time from git history,
          // which is more reliable than file system creation time under most CI environments.
          return getGitDate("created", entry.src) ||
            getGitDate("modified", entry.src) || info.birthtime ||
            new Date();
      }
    }

    try {
      return parseDate(date);
    } catch {
      throw new Error(`Invalid date: ${date} (${entry?.src})`);
    }
  }

  return info?.birthtime || info?.mtime || new Date();
}
