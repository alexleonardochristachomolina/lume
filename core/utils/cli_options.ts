import { parseArgs } from "../../deps/cli.ts";
import type { DeepPartial } from "./object.ts";
import type { SiteOptions } from "../site.ts";

export function getOptionsFromCli(
  options: DeepPartial<SiteOptions>,
): DeepPartial<SiteOptions> {
  const cli = parseArgs(Deno.args, {
    string: ["src", "dest", "location", "port"],
    boolean: ["serve", "open"],
    alias: { dev: "d", serve: "s", port: "p", open: "o" },
    ["--"]: true,
  });

  if (cli.src) {
    options.src = cli.src;
  }

  if (cli.dest) {
    options.dest = cli.dest;
  }

  const serveMode = cli.serve || cli._[0] === "cms";

  if (cli.port) {
    (options.server ||= {}).port = parseInt(cli.port);
  } else if (serveMode) {
    (options.server ||= {}).port = 3000;
  }

  let location: URL;

  if (cli.location) {
    location = new URL(cli.location);
  } else if (options.location && !serveMode) {
    location = options.location as URL;
  } else {
    location = new URL("http://localhost");
  }

  let port: number;

  if (cli.port) {
    port = parseInt(cli.port);
  } else if (location.port) {
    port = parseInt(location.port);
  } else if (options.server?.port) {
    port = options.server.port;
  } else {
    port = cli.serve ? 3000 : location.protocol === "https:" ? 443 : 80;
  }

  (options.server ||= {}).port = port;
  location.port = port.toString();
  options.location = location;

  if (cli.open) {
    (options.server ||= {}).open = cli.open;
  }

  return options;
}
