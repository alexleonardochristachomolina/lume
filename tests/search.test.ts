import { build, getSite } from "./utils.ts";
import { assertSnapshot } from "../deps/snapshot.ts";
import { buildFilter, buildSort } from "../core/searcher.ts";

Deno.test("Search by Tags", async (t) => {
  const filter = buildFilter("foo bar");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search by NOT Tags", async (t) => {
  const filter = buildFilter("foo !bar");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search by Equal", async (t) => {
  const filter = buildFilter("foo=bar");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search by Equal undefined", async (t) => {
  const filter = buildFilter("foo=undefined");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search by Equal null", async (t) => {
  const filter = buildFilter("foo=null");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search by Upper than", async (t) => {
  const filter = buildFilter("foo>bar");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search by Upper or equals than", async (t) => {
  const filter = buildFilter("foo>=bar");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search by Lower than", async (t) => {
  const filter = buildFilter("foo<bar");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search by Lower or equals than", async (t) => {
  const filter = buildFilter("foo<=bar");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search by Not Equal", async (t) => {
  const filter = buildFilter("foo!=bar");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search by Not Equal alt", async (t) => {
  const filter = buildFilter("!foo=bar");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search by Starts With", async (t) => {
  const filter = buildFilter("foo^=bar");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search by NOT Starts With", async (t) => {
  const filter = buildFilter("!foo^=bar");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search by Ends With", async (t) => {
  const filter = buildFilter("foo$=bar");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search by Contains", async (t) => {
  const filter = buildFilter("foo*=bar");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search by Tags with OR", async (t) => {
  const filter = buildFilter("foo|bar");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search by Equal with OR", async (t) => {
  const filter = buildFilter("foo=bar|baz");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search by Not Equal with OR", async (t) => {
  const filter = buildFilter("foo!=bar|baz");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search by Starts With with OR", async (t) => {
  const filter = buildFilter("foo^=bar|baz");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search by Ends With with OR", async (t) => {
  const filter = buildFilter("foo$=bar|baz");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search by Contains with OR", async (t) => {
  const filter = buildFilter("foo*=bar|baz");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search Date by Equal", async (t) => {
  const filter = buildFilter("foo=2000-01-02");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search Date by Not Equal", async (t) => {
  const filter = buildFilter("foo!=2000-01-02");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search Date by lower than", async (t) => {
  const filter = buildFilter("foo<2000-01-02T18:00");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search Date by lower or equals than", async (t) => {
  const filter = buildFilter("foo<=2000-01-02T18:00");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search Date by upper than", async (t) => {
  const filter = buildFilter("foo>2000-01-02T18:00");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Search Date by upper or equals than", async (t) => {
  const filter = buildFilter("foo>=2000-01-02T18:00");

  await assertSnapshot(t, filter?.toString());
});

Deno.test("Sort by one field", async (t) => {
  const sort = buildSort("order");

  await assertSnapshot(t, sort);
});

Deno.test("Sort by one field DESC", async (t) => {
  const sort = buildSort("order=desc");

  await assertSnapshot(t, sort);
});

Deno.test("Sort by two fields", async (t) => {
  const sort = buildSort("order title");

  await assertSnapshot(t, sort);
});

Deno.test("Sort by two fields, second is DESC", async (t) => {
  const sort = buildSort("order title=desc");

  await assertSnapshot(t, sort);
});

Deno.test("Sort by two locale fields", async (t) => {
  const sort = buildSort("order=asc-locale title=desc-locale");

  await assertSnapshot(t, sort.toString());
});

Deno.test("Test the file searcher", async (t) => {
  const site = getSite({
    src: "normal",
  });

  site.loadAssets([".css"]);
  site.copy([".png"]);

  await build(site);

  const { search } = site;

  await assertSnapshot(t, search.files().sort());
  await assertSnapshot(t, search.files("/*.png").sort());
  await assertSnapshot(t, search.files(/\.png$/).sort());
});
