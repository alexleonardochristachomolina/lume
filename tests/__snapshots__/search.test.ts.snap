export const snapshot = {};

snapshot[`Search by Tags 1`] = `"(data) => data.tags?.includes(value0) && data.tags?.includes(value1)"`;

snapshot[`Search by NOT Tags 1`] = `"(data) => data.tags?.includes(value0) && !data.tags?.includes(value1)"`;

snapshot[`Search by Equal 1`] = `"(data) => data.foo === value0"`;

snapshot[`Search by Equal undefined 1`] = `"(data) => data.foo === value0"`;

snapshot[`Search by Equal null 1`] = `"(data) => data.foo === value0"`;

snapshot[`Search by Upper than 1`] = `"(data) => data.foo > value0"`;

snapshot[`Search by Upper or equals than 1`] = `"(data) => data.foo >= value0"`;

snapshot[`Search by Lower than 1`] = `"(data) => data.foo < value0"`;

snapshot[`Search by Lower or equals than 1`] = `"(data) => data.foo <= value0"`;

snapshot[`Search by Not Equal 1`] = `"(data) => data.foo !== value0"`;

snapshot[`Search by Not Equal alt 1`] = `"(data) => data.foo !== value0"`;

snapshot[`Search by Starts With 1`] = `"(data) => data.foo?.startsWith(value0)"`;

snapshot[`Search by NOT Starts With 1`] = `"(data) => !data.foo?.startsWith(value0)"`;

snapshot[`Search by Ends With 1`] = `"(data) => data.foo?.endsWith(value0)"`;

snapshot[`Search by Contains 1`] = `"(data) => data.foo?.includes(value0)"`;

snapshot[`Search by Tags with OR 1`] = `"(data) => value0.some((i) => data.tags?.includes(i))"`;

snapshot[`Search by Equal with OR 1`] = `"(data) => value0.some((i) => data.foo === i)"`;

snapshot[`Search by Not Equal with OR 1`] = `"(data) => value0.some((i) => data.foo !== i)"`;

snapshot[`Search by Starts With with OR 1`] = `"(data) => value0.some((i) => data.foo?.startsWith(i))"`;

snapshot[`Search by Ends With with OR 1`] = `"(data) => value0.some((i) => data.foo?.endsWith(i))"`;

snapshot[`Search by Contains with OR 1`] = `"(data) => value0.some((i) => data.foo?.includes(i))"`;

snapshot[`Search Date by Equal 1`] = `"(data) => data.foo?.getTime() === value0.getTime()"`;

snapshot[`Search Date by Not Equal 1`] = `"(data) => data.foo?.getTime() !== value0.getTime()"`;

snapshot[`Search Date by lower than 1`] = `"(data) => data.foo?.getTime() < value0.getTime()"`;

snapshot[`Search Date by lower or equals than 1`] = `"(data) => data.foo?.getTime() <= value0.getTime()"`;

snapshot[`Search Date by upper than 1`] = `"(data) => data.foo?.getTime() > value0.getTime()"`;

snapshot[`Search Date by upper or equals than 1`] = `"(data) => data.foo?.getTime() >= value0.getTime()"`;

snapshot[`Sort by one field 1`] = `[Function (anonymous)]`;

snapshot[`Sort by one field DESC 1`] = `[Function (anonymous)]`;

snapshot[`Sort by two fields 1`] = `[Function (anonymous)]`;

snapshot[`Sort by two fields, second is DESC 1`] = `[Function (anonymous)]`;

snapshot[`Sort by two locale fields 1`] = `'function (a, b) { return (a.order == b.order ? (a.title == b.title ? 0 : collator.compare(b.title || "", a.title || "")) : collator.compare(a.order || "", b.order || "")); }'`;

snapshot[`Test the file searcher 1`] = `
[
  "/404.html",
  "/favicon.png",
  "/overrided-page2/index.html",
  "/page5/index.html",
  "/page_3",
  "/pages/new-name/page7/index.html",
  "/pages/page4/index.html",
  "/pages/page6/index.html",
  "/static/index.html",
  "/styles.css",
]
`;

snapshot[`Test the file searcher 2`] = `
[
  "/favicon.png",
]
`;

snapshot[`Test the file searcher 3`] = `
[
  "/favicon.png",
]
`;
